### PR TITLE
Host configs for new LLNL machines + cuda12

### PR DESCRIPTION
**Pull request by victorapm**
Add new host configs:

- lassen - cuda 12
- ruby - gcc-12 and clang-14
- dane - gcc-12

See https://github.com/GEOS-DEV/LvArray/pull/319 and https://github.com/GEOS-DEV/thirdPartyLibs/pull/270

### CODE REVIEW

**Comment by TotoGaz** [comment]
@sframba Do you know if CUDA 12 is supported on Pangea3
@matteofrigo5 Same question on Sherlock

File: host-configs/LLNL/lassen-clang-13-cuda-12.cmake [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+set(MPI_Fortran_COMPILER /usr/tce/packages/spectrum-mpi/spectrum-mpi-rolling-release-gcc-8.3.1/bin/mpifort CACHE PATH "")
+
+include(${CMAKE_CURRENT_LIST_DIR}/lassen-base.cmake)
+
+# Overwrite options set on lassen-base.cmake
+set(ENABLE_OPENMP OFF CACHE BOOL "" FORCE)
+set(ENABLE_CUDA_NVTOOLSEXT ON CACHE BOOL "")
+
+# Overwrite ESSL defaults from lassen-base.cmake
+# Reason: libesslsmpcuda.so depends on cuda-11
```
Comment (CusiniM) [inline host-configs/LLNL/lassen-clang-13-cuda-12.cmake:21]: so we don't really need it?

Comment (victorapm) [reply host-configs/LLNL/lassen-clang-13-cuda-12.cmake:21]: `libesslsmpcuda.so`? Yes, geos doesn't complain if I remove it. It seems it's not being used anywhere

**Comment by sframba** [comment]
> @sframba Do you know if CUDA 12 is supported on Pangea3 @matteofrigo5 Same question on Sherlock

Not sure, we'll have to ask the IBM support. We have upgraded to `cuda 11.5.0` at the end of 2023. Is cuda 12 needed to solve the `externalSolvers` unit test issue? Or to access some new `hypre` features on GPU?

**Comment by matteofrigo5** [comment]
I don't see any constraints regarding Sherlock. We have these four versions of CUDA 12 installed: (12.0.0 12.1.1 12.2.0 12.4.0). @victorapm, do you suggest any one in particular?

**Comment by TotoGaz** [comment]
> Not sure, we'll have to ask the IBM support. We have upgraded to `cuda 11.5.0` at the end of 2023. Is cuda 12 needed to solve the `externalSolvers` unit test issue? Or to access some new `hypre` features on GPU?

It's mainly for [this](https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/).

**Review by CusiniM — APPROVED** [review]

**Comment by CusiniM** [comment]
@rrsettgast @wrtobin @castelletto1  I have installed the tpls for they ruby builds and for the lassen cuda-12 one. 

On `ruby`:

- clang build works perfectly fine (only tested with `make`)  and both unit (apart from the doxygen one) and intergratedTests pass.
- gcc build works fine to build geos but there is a linking error in the unit tests (see below).

````
/usr/tce/packages/gcc/gcc-12.1.1/bin/ld: /usr/gapps/GEOSX/thirdPartyLibs/2024-06-19/install-ruby-gcc-12-release/trilinos/lib/libtpetra.so.13.4.1: undefined reference to `KokkosSparse::Impl::SPMV<double const, int const, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u>, int const, double const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<3u>, double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u>, true, true>::spmv(KokkosKernels::Experimental::Controls const&, char const*, double const&, KokkosSparse::CrsMatrix<double const, int const, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u>, int const> const&, Kokkos::View<double const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<3u> > const&, double const&, Kokkos::View<double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&)' 
````

we are on an oldish version of Trilinos (13.4.1) which is from 2022. We could try to move the latest release (https://github.com/trilinos/Trilinos/releases) which is from a couple of months ago and see if that fixes the issue.


- I have not tested the new `lassen` build with cuda-12 yet but I think @victorapm did and should be working.

I don't have a bank on `dane` yet so I can't do much there but I am pretty sure that we can get everything to work on that system too. 

Anyways, since we have 1 fully working build, If you are okay with it, I think we can merge this and abandon quartz in favor of ruby.

**Comment by victorapm** [comment]
> I have not tested the new lassen build with cuda-12 yet but I think @victorapm did and should be working.

After some trial and error with different compilers, it works!

https://github.com/GEOS-DEV/LvArray/blob/fe2ad691194321af4f5f9cab3593016d8d0fc645/host-configs/LLNL/lassen-clang-13-cuda-12.cmake#L3-L10

PS: the build on Dane also works fine (tested it with a couple of simulations)

File: .github/workflows/ci_tests.yml [hunk 2]
```diff
@@ -284,6 +284,16 @@ jobs:
...
             DOCKER_CERTS_DIR: "/etc/pki/ca-trust/source/anchors"
             DOCKER_CERTS_UPDATE_COMMAND: "update-ca-trust"
+          
+          - name: Rockylinux (8, clang 17.0.6, cuda 12.5)
+            BUILD_AND_TEST_CLI_ARGS: "--no-run-unit-tests --no-install-schema"
+            CMAKE_BUILD_TYPE: Release
+            DOCKER_REPOSITORY: geosx/rockylinux8-clang17-cuda12.5
+            RUNS_ON: streak2
+            NPROC: 16
+            DOCKER_RUN_ARGS: "--cpus=16 --memory=256g --runtime=nvidia -v /etc/pki/ca-trust/source/anchors/:/etc/pki/ca-trust/source/anchors/llnl:ro"
```
Comment (rrsettgast) [inline .github/workflows/ci_tests.yml]:
```suggestion
            NPROC: 8
            DOCKER_RUN_ARGS: "--cpus=8 --memory=128g --runtime=nvidia -v /etc/pki/ca-trust/source/anchors/:/etc/pki/ca-trust/source/anchors/llnl:ro"
```

**Review by rrsettgast — APPROVED** [review]
@CusiniM aren't dane and ruby the same stack? Can we merge the files?

**Comment by CusiniM** [comment]
> @CusiniM aren't dane and ruby the same stack? Can we merge the files?

yeah, that's why I created that `llnl-cpu-base.cmake`. I think it's convenient to keep the file separate if we ever  want/need to customize something but yeah, they should be identical AFAIK. I would also expect binaries to work for both systems. The only difference would be the 1 ats parameter.  I don't have a bank on dane yet so I have not been able to do much testing there.

**Comment by CusiniM** [comment]
@sframba can you guys have a look at why [this](https://github.com/GEOS-DEV/GEOS/actions/runs/9684740474/job/26724623550?pr=3067#step:8:3313) cuda-12 job is failing when building `ElasticFirstOrderWaveEquationSEM` ?

**Comment by CusiniM** [comment]
@sframba can you guys have a look at why [this](https://github.com/GEOS-DEV/GEOS/actions/runs/9685520478/job/26725810422?pr=3067) cuda-12 build is crashing when building `ElasticFirstOrderWaveEquationSEM`?

File: .github/workflows/ci_tests.yml [hunk 3]
```diff
@@ -284,6 +284,16 @@ jobs:
...
             DOCKER_CERTS_DIR: "/etc/pki/ca-trust/source/anchors"
             DOCKER_CERTS_UPDATE_COMMAND: "update-ca-trust"
+          
+          - name: Rockylinux (8, clang 17.0.6, cuda 12.5)
+            BUILD_AND_TEST_CLI_ARGS: "--no-run-unit-tests --no-install-schema"
+            CMAKE_BUILD_TYPE: Release
+            DOCKER_REPOSITORY: geosx/rockylinux8-clang17-cuda12.5
+            RUNS_ON: streak2
+            NPROC: 2
+            DOCKER_RUN_ARGS: "--cpus=2 --memory=128g --runtime=nvidia -v /etc/pki/ca-trust/source/anchors/:/etc/pki/ca-trust/source/anchors/llnl:ro"
```
Comment (rrsettgast) [inline .github/workflows/ci_tests.yml]:
```suggestion
            DOCKER_RUN_ARGS: "--cpus=1 --memory=128g --runtime=nvidia -v /etc/pki/ca-trust/source/anchors/:/etc/pki/ca-trust/source/anchors/llnl:ro"
```

**Pull request merged**